kafka-client/src/main/java/dev/responsive/kafka/api/ResponsiveKafkaStreams.java

```diff
@@ -417,4 +356,93 @@ public boolean close(final CloseOptions options) {
     closeInternal();
     return closed;
   }
+
+  protected static class Params {
+
+    final Topology topology;
+    final ResponsiveConfig responsiveConfig;
+    final StreamsConfig streamsConfig;
+    final ResponsiveMetrics metrics;
+    final ResponsiveStoreRegistry storeRegistry;
+
+    // can be set during construction
+    Time time = Time.SYSTEM;
+    KafkaClientSupplier clientSupplier = new DefaultKafkaClientSupplier();
+    CassandraClientFactory cassandraFactory = new DefaultCassandraClientFactory();
+
+    // initialized on init()
+    SharedClients sharedClients;
+    ResponsiveKafkaClientSupplier responsiveKafkaClientSupplier;
+
+    public Params(final Topology topology, final Map<?, ?> configs) {
+      this.topology = topology;
+
+      // this should be the only place this is logged!
+      this.responsiveConfig = ResponsiveConfig.loggedConfig(configs);
+      this.streamsConfig = ResponsiveStreamsConfig.streamsConfig(configs);
+
+      this.metrics = createMetrics(streamsConfig);
+      this.storeRegistry = new ResponsiveStoreRegistry();
+    }
+
+    public Params withClientSupplier(final KafkaClientSupplier clientSupplier) {
+      this.clientSupplier = clientSupplier;
+      return this;
+    }
+
+    public Params withCassandraClientFactory(final CassandraClientFactory clientFactory) {
+      this.cassandraFactory = clientFactory;
+      return this;
+    }
+
+    public Params withTime(final Time time) {
+      this.time = time;
+      return this;
+    }
+
+    // we may want to consider wrapping this in an additional Builder to ensure
+    // that it's impossible to use a Params instance that hasn't called build(),
+    // but that felt a little extra
+    public Params build() {
+      this.responsiveKafkaClientSupplier = new ResponsiveKafkaClientSupplier(
+          clientSupplier,
+          streamsConfig,
+          storeRegistry,
+          metrics
+      );
+      final var admin = responsiveKafkaClientSupplier.getAdmin(responsiveConfig.originals());
+
+      final var backendType = StorageBackend.valueOf(responsiveConfig
+          .getString(STORAGE_BACKEND_TYPE_CONFIG)
+          .toUpperCase(Locale.ROOT)
+      );
+
+      switch (backendType) {
+        case CASSANDRA:
+          final var cqlSession = cassandraFactory.createCqlSession(responsiveConfig);
```

Comment:
yes! I was imagining something pretty similar, but didn't want it to spiral at this point. all of this should definitely be encapsulated away somewhere so that we can also more easily test different implementations instead of repeating this same logic throughout